File: editor/en/[1289]Minimum Falling Path Sum II.java
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int minFallingPathSum(int[][] grid) {
        int n = grid.length;

        for (int i = 1; i < n; i++) {
            int min1 = Integer.MAX_VALUE;
            int min2 = Integer.MAX_VALUE;

            for (int j = 0; j < n; j++) {
                int value = grid[i - 1][j];

                if (value < min1) {
                    min2 = min1;
                    min1 = value;
                } else if (value < min2) {
                    min2 = value;
                }
            }

            for (int j = 0; j < n; j++) {
                if (grid[i - 1][j] == min1) {
                    grid[i][j] += min2;
                } else {
                    grid[i][j] += min1;
                }
            }
        }

        int minSum = Integer.MAX_VALUE;

        for (int i = 0; i < n; i++) {
            minSum = Math.min(minSum, grid[n - 1][i]);
        }

        return minSum;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[1289]Minimum Falling Path Sum II.py
# leetcode submit region begin(Prohibit modification and deletion)
import heapq


class Solution:
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        n = len(grid)

        for i in range(1, n):
            two_smallest = heapq.nsmallest(2, grid[i - 1])

            for j in range(n):
                grid[i][j] += two_smallest[1] if grid[i - 1][j] == two_smallest[0] else two_smallest[0]

        return min(grid[-1])
    # leetcode submit region end(Prohibit modification and deletion)
